engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala

```diff
@@ -268,11 +285,14 @@ case object DataflowAnalysis extends IRPass {
         info.dependents.updateAt(nameDep, Set(bindingDep))
         info.dependencies.updateAt(bindingDep, Set(expressionDep, nameDep))
 
-        binding
-          .copy(
-            name       = name.updateMetadata(new MetadataPair(this, info)),
-            expression = analyseExpression(expression, info)
-          )
+        val expression1 = analyseExpression(expression, info)
+        name.updateMetadata(new MetadataPair(this, info))
+        (if (expression1 != expression)
+           binding
+             .copy(
```

Comment:
When I heard about this effort, I thought we'll do the change inside of `copy`. If all the values are `==` then just return `this`...

Reply:
I did consider it, but I don't think it will place nice with `duplicate` as it also calls `copy`

Reply:
Typical pattern in `duplicate` is:
```scala
passData =
          if (keepMetadata) passData.duplicate else new MetadataStorage()
```
if you guarantee that `passData.duplicate` **always creates new** instance. Then `duplicate` will always _differ at least in one field_ and `copy` would really create a fresh copy.

Reply:
>>I thought we'll do the change inside of `copy`. If all the values are `==` then just return `this`...
>  I don't think it will place nice with `duplicate`

The problem with doing these comparisions outside of `copy` is that we need to think about them when writing new code like #10729 ...